Final compile fixes for pathfinder

fschutt opened the pull request:
Closes #2645. 

Note: On Windows, pathfinder won't yet run properly, partly due to https://github.com/pcwalton/pathfinder/pull/82 and partly due to some `RenderTask`-related bugs (currently it crashes with "unexpected render task: glyph task") in ` webrender::render_task::RenderTask::uv_rect_kind`.

However, pathfinder now at least compiles on Windows, Linux and Mac. Also added various `#[derive(Debug)]` necessary for debugging the pathfinder crashes.

glennw commented:
r? @pcwalton

kvark commented:
Can we have that path tested on CI?

fschutt commented:
Yes although, as I said, it will crash at runtime. Compilation tests should work, though.

kvark commented:
@fschutt I think we should keep Travis is sync with TaskCluster in what it's building for as long as Travis is useful, to avoid confusion at least.

kvark commented on webrender/src/glyph_rasterizer.rs:
```diff
@@ -8,79 +8,68 @@ use api::{ColorF, ColorU};
+        use std::ops::Deref;
+        use tiling::RenderTargetKind;
+        use webrender_api::{DeviceIntPoint, DevicePixel};
+    } else if #[cfg(not(feature = "pathfinder"))] {
```
does this need `if` after `else`?

kvark commented on webrender/src/glyph_rasterizer.rs:
```diff
@@ -8,79 +8,68 @@ use api::{ColorF, ColorU};
 use tiling::SpecialRenderPasses;
-#[cfg(feature = "pathfinder")]
-use webrender_api::{DeviceIntPoint, DevicePixel};
+cfg_if! {
```
The amount of diversion between the pathfinder and whatnot features of this module is pretty high. We should consider splitting those into separate modules, potentially leaving something shared in another module. This would be cleaner than multiple `#[cfg(feature = "pathfinder")]` spread over the code.

fschutt replied:
Yeah ok, but this is going to take a while to split up the files and resolve all the errors. I'll do two seperate files for the pathfinder and the no_pathfinder configuration.

fschutt replied:
btw, is there any difference between `use api::` and `use webrender_api::` or are they the same?

kvark replied:
No, this is because of this line in "lib.rs": `pub use webrender_api as api;`
I'd much prefer us just doing `pub extern crate webrender_api as api;` in the first place, without any ambiguity.

pcwalton commented on webrender/src/glyph_rasterizer/glyph_rasterizer_ext_pathfinder.rs:
```diff
@@ -0,0 +1,314 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Module only available when pathfinder is activated
```
How about just naming this `pathfinder.rs`? It's shorter :)

fschutt commented:
@kvark @pcwalton All done, tests passed. Ready to merge.

kvark commented:
Reviewed 2 of 4 files at r1, 3 of 3 files at r2, 2 of 3 files at r3, 1 of 6 files at r4, 3 of 6 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[appveyor.yml, line 23 at r5](https://reviewable.io/reviews/servo/webrender/2723#-LBg5IYn37UVm3RffWIz:-LBg5IYn37UVm3RffWJ-:b2a0hnd) ([raw file](https://github.com/servo/webrender/blob/465bfdf3424f85592d24e38063cd429499356d62/appveyor.yml#L23)):*
> ```YAML
>   - cd ../webrender
>   - cargo test --verbose
>   - cargo build --verbose --no-default-features --features pathfinder
> ```

somewhat less relevant, I wonder if we can use `cargo check` instead of `cargo build` on CI to speed it up a bit

---

*[webrender/src/glyph_rasterizer/no_pathfinder.rs, line 24 at r5](https://reviewable.io/reviews/servo/webrender/2723#-LBg5o-l5ksrGyC4Hihf:-LBg5o-l5ksrGyC4Hihg:byiqzn7) ([raw file](https://github.com/servo/webrender/blob/465bfdf3424f85592d24e38063cd429499356d62/webrender/src/glyph_rasterizer/no_pathfinder.rs#L24)):*
> ```Rust
> use std::collections::hash_map::Entry;
> 
> pub(in super) fn create_pathfinder_font_context() -> Result<(), ResourceCacheError> {
> ```

why do we need pathfinder-related methods in this module at all?

---


*Comments from [Reviewable](https://reviewable.io/reviews/servo/webrender/2723)*
<!-- Sent from Reviewable.io -->

fschutt commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[appveyor.yml, line 23 at r5](https://reviewable.io/reviews/servo/webrender/2723#-LBg5IYn37UVm3RffWIz:-LBjr8WJ9CaprV6QoGS0:bri26f7) ([raw file](https://github.com/servo/webrender/blob/465bfdf3424f85592d24e38063cd429499356d62/appveyor.yml#L23)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

somewhat less relevant, I wonder if we can use `cargo check` instead of `cargo build` on CI to speed it up a bit
</blockquote></details>

I think we can change from "cargo build" to "cargo check" whenever we are not doing anything with the final binary. I.e. we just see if it compiles, then throw the binary away. I am not aware of checks where you have to compile the crate in order to verify that it compiles. I'll change it for all of the CI.

---

*[webrender/src/glyph_rasterizer/no_pathfinder.rs, line 24 at r5](https://reviewable.io/reviews/servo/webrender/2723#-LBg5o-l5ksrGyC4Hihf:-LBjrysx1HWuwHQDLldv:b4t78or) ([raw file](https://github.com/servo/webrender/blob/465bfdf3424f85592d24e38063cd429499356d62/webrender/src/glyph_rasterizer/no_pathfinder.rs#L24)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

why do we need pathfinder-related methods in this module at all?
</blockquote></details>

The `FontContexts` struct currently contains a pathfinder_context field, although that is unused on non-pathfinder builds. I'll change that, thanks.

---


*Comments from [Reviewable](https://reviewable.io/reviews/servo/webrender/2723)*
<!-- Sent from Reviewable.io -->

fschutt commented:
@kvark Just notifying you that the issues are fixed now. I changed `cargo build` to `cargo check` and removed the pathfinder-related functions from the `no_pathfinder` module.

kvark commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/servo/webrender/2723)*
<!-- Sent from Reviewable.io -->

kvark commented:
As this is just moving things around, I don't expect any reftest failures, but we'll stay on guard.
Thank you for pushing this to the end!
@bors-servo r+

bors-servo commented:
:pushpin: Commit f06f503 has been approved by `kvark`

<!-- @bors-servo r=kvark f06f5030be97826cf235311bf35b557bf61d2788 -->

bors-servo commented:
:hourglass: Testing commit f06f5030be97826cf235311bf35b557bf61d2788 with merge 1eb46da9a8ff5fc092834b08a1578cb636cac560...
